mvebu: fix LEDs on Fortinet FortiGate devices and IIJ SA-W2

musashino205 opened the pull request:
This patch series fixes LEDs on Fortinet FortiGate 30E/50E and IIJ SA-W2.

related issue: https://github.com/openwrt/openwrt/issues/14628

cc: @Ansuel

musashino205 committed:
mvebu: fix LEDs on Fortinet FortiGate devices

On Fortinet FortiGate 30E/50E, some multiple LEDs have no "function"
property and only "color" property is available for the new binding of
LED on Linux Kernel.
9d93b6d091 ("mvebu: drop redundant label with new LED color/function
format") removes "label" property from LEDs, then, multiple "<color>:"
(ex.: "green:"/"red:"/"amber:") will be appeared as LED names and
renamed to "<color>:_<num>" (ex.: "green:_1", "green:_2", ...) by
kernel.

log:

[   12.425170] leds-gpio gpio-leds: Led green: renamed to green:_1 due to name collision
[   12.520390] leds-gpio gpio-leds: Led amber: renamed to amber:_1 due to name collision
[   12.614931] leds-gpio gpio-leds: Led green: renamed to green:_2 due to name collision
[   12.709895] leds-gpio gpio-leds: Led green: renamed to green:_3 due to name collision
[   12.804439] leds-gpio gpio-leds: Led amber: renamed to amber:_2 due to name collision
[   12.898969] leds-gpio gpio-leds: Led green: renamed to green:_4 due to name collision
[   12.993504] leds-gpio gpio-leds: Led amber: renamed to amber:_3 due to name collision
[   13.088033] leds-gpio gpio-leds: Led green: renamed to green:_5 due to name collision
[   13.182570] leds-gpio gpio-leds: Led green: renamed to green:_6 due to name collision
[   13.277103] leds-gpio gpio-leds: Led amber: renamed to amber:_4 due to name collision
[   13.371636] leds-gpio gpio-leds: Led green: renamed to green:_7 due to name collision

/sys/class/leds:

root@OpenWrt:/# ls /sys/class/leds/
amber:        amber:_4      green:_2      green:_6      red:alarm
amber:_1      amber:alarm   green:_3      green:_7      red:status
amber:_2      green:        green:_4      green:status
amber:_3      green:_1      green:_5      red:

Fix this issue by adding missing "function" (and "function-enumerator")
property those to LEDs on Fortinet FortiGate devices.
Note: there is no appropriate function for "ha" LEDs in
dt-bindings/leds/common.h, so use the hardcoded string for them instead.

Fixes: 9d93b6d091 ("mvebu: drop redundant label with new LED color/function format")
Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 committed:
mvebu: fix LEDs on IIJ SA-W2

On IIJ SA-W2, some multiple LEDs have no "function" property and only
"color" property is available for the newer binding of LED on Linux
Kernel.
9d93b6d091 ("mvebu: drop redundant label with new LED color/function
format") removes "label" property from LEDs, then, multiple "<color>:"
(ex.: "green:"/"red:") will be appeared and renamed to "<color>:_<num>"
(ex.: "green:_1", "green:_2", ...) by kernel.

log:

[    1.911118] leds-gpio leds: Led green: renamed to green:_1 due to name collision
[    1.918600] leds-gpio leds: Led red: renamed to red:_1 due to name collision
[    1.925727] leds-gpio leds: Led green: renamed to green:_2 due to name collision
[    1.933202] leds-gpio leds: Led red: renamed to red:_2 due to name collision
[    1.940321] leds-gpio leds: Led green: renamed to green:_3 due to name collision
[    1.947797] leds-gpio leds: Led red: renamed to red:_3 due to name collision
[    1.954939] leds-gpio leds: Led green: renamed to green:_4 due to name collision
[    1.962456] leds-gpio leds: Led green: renamed to green:_5 due to name collision

/sys/class/leds:

root@OpenWrt:/# ls /sys/class/leds/
green:        green:_3      green:status  red:_2
green:_1      green:_4      red:          red:_3
green:_2      green:_5      red:_1        red:status

Fix this issue by adding missing "function" (and "function-enumerator")
property to those LEDs on IIJ SA-W2.

Fixes: 9d93b6d091 ("mvebu: drop redundant label with new LED color/function format")
Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

namiltd commented on target/linux/mvebu/files/arch/arm/boot/dts/armada-380-iij-sa-w2.dts:
```diff
@@ -66,12 +66,14 @@
 		pinctrl-0 = <&pmx_leds_pins>;
 
 		led-0 {
+			label = "green:wlan5g";
```
```
function = LED_FUNCTION_WLAN;
function-enumerator = <5>;
```
and similarly further

musashino205 replied:
I thought about using them once, but it might be confusing for users without the suffix 'g', so I used label instead.

stklcode replied:
We already have mixed variants across the tree.
0/1 (matching the interface names), 2/5, 24/50 and labels with “g“.

I personally do have a light preference  for using color/function for generic names, but I do see valid arguments for all versions. If frequencies are fixed on the radios, I’d be fine with either. Otherwise I’d use 0/1/.. and just map the label to the radio name.

hauke replied:
@rmilecki added separate LED functions for 2.4, 5 and 6 GHz Wifi to upstream kernel, see:
https://github.com/torvalds/linux/commit/ec18a2a83b8b9f7e39c80105ea148c769c46227b

namiltd reviewed (commented):
Use "function" instead of "label"

musashino205 commented:
UPDATE:

- convert WLAN LEDs on SA-W2 to color/function

sppmasterspp commented:
@musashino205 
Can you fix the LED bug for qnap 301w either?
Thanks.

> This patch series fixes LEDs on Fortinet FortiGate 30E/50E and IIJ SA-W2.
> 
> related issue: #14628
> 
> cc: @Ansuel

musashino205 commented:
UPDATE:

- convert all LEDs to color/function
- rebase from current main branch

hauke commented on target/linux/mvebu/files/arch/arm/boot/dts/armada-380-iij-sa-w2.dts:
```diff
@@ -91,44 +95,56 @@
 		led-4 {
 			gpios = <&gpio1 3 GPIO_ACTIVE_HIGH>;
 			color = <LED_COLOR_ID_GREEN>;
+			function = "mobile";
```
I do not like it when this is defined as a free text, maybe try to get such a constant into the upstream kernel. an LED for mobile network connection is pretty common.

musashino205 replied:
I see. I'll try to send a patch for adding "mobile".

rmilecki commented on target/linux/mvebu/files/arch/arm/boot/dts/armada-380-iij-sa-w2.dts:
```diff
@@ -68,12 +68,16 @@
 			gpios = <&gpio0 19 GPIO_ACTIVE_HIGH>;
 			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WLAN;
+			function-enumerator = <5>;
```
I think you can now use `LED_FUNCTION_WLAN_5GHZ`

rmilecki commented on target/linux/mvebu/files/arch/arm/boot/dts/armada-380-iij-sa-w2.dts:
```diff
@@ -91,44 +95,56 @@
 			gpios = <&gpio1 12 GPIO_ACTIVE_HIGH>;
 			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_WLAN;
+			function-enumerator = <2>;
```
and `LED_FUNCTION_WLAN_2GHZ`

musashino205 committed:
kernel: add pending patches for new LED_FUNCTION_* definitions

Add pending patches to add LED_FUNCTION_MOBILE and LED_FUNCTION_SPEED_*
definitions for Fortinet FortiGate devices and IIJ SA-W2.

Signed-off-by: INAGAKI Hiroshi <musashino.open@gmail.com>

musashino205 commented:
UPDATE:

- add pending patches for ```LED_FUNCTION_MOBILE``` and `LED_FUNCTION_SPEED_*`
- use `LED_FUNCTION_SPEED_*` on FortiGate 30E/50E
- use `LED_FUNCTION_MOBILE` and `LED_FUNCTION_WLAN_nGHZ` on IIJ SA-W2

musashino205 commented:
UPDATE:

- rebase from current main branch
  - apply this fix for Liniux Kernel 6.6 as well

musashino205 commented:
UPDATE:

- rebase from current main branch
